sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/constraints.scala

```diff
@@ -0,0 +1,179 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql.catalyst.expressions
+
+import org.antlr.v4.runtime.ParserRuleContext
+
+import org.apache.spark.sql.catalyst.parser.ParseException
+import org.apache.spark.sql.types.{DataType, StringType}
+
+trait TableConstraint {
+
+  /** Returns the name of the constraint */
+  def name: String
+
+  /** Returns the characteristics of the constraint (e.g., ENFORCED, RELY) */
+  def characteristic: ConstraintCharacteristic
+
+  /** Creates a new constraint with the given name
+   * @param name Constraint name
```

Comment:
Question: Do we need an extra empty line before params? What's the style preference in Spark?